Fix --watch mode

nikuda opened the pull request:
Using `addOutputFileArg` to pass the output string to `addSystemCommand` fixes `compile` fn in `zig build --watch` mode.

Having access to return value of `addOutputFileArg` allows you create a module out of the output

```
const result = try compile(b, .{
    .dep_shdc = b.dependency("shdc", .{}),
    .input = input_file_path,
    .output = output_file_path,
    .slang = .{
        .glsl430 = true,
        .metal_macos = true,
        .hlsl5 = true,
    },
    .reflection = true,
});

const run_shdc = result.run;

// Register the generated file as a module
const shader_module = b.addModule(import_file_path, .{ .root_source_file = result.output });
shader_module.addImport("sokol", sokol_module);
exe.root_module.addImport(import_file_path, shader_module);
```

So the flow is, by having `--watch` fixed, files are no longer copied relatively but to `.zig-cache` - to use them from there one method is to make them into a module which works well, but do you then:

- Create another `compileTo` method that copies the result.output file to a relative location, preserving existing behaviour (while also working with --watch) or leave that for user end?
- Also create a registerModule fn to handling compiling and registering module as a convenience function too?

nikuda committed:
Fix --watch mode

floooh commented:
I like the change, but we should optionally have a mode where the output file can be in the source tree and directly imported (e.g. similar to how it behaves now)...

In the cmake-based projects I have similar 'dual behaviour':

- in the sokol-samples, the output files are written to the cmake build directory (e.g. what cmake calls "out of source"), and the output files are never committed into version control
- ...and other projects generate the output file into the source tree, and in this case I generally commit the output files into version control (e.g. you have both `shader.glsl` and `shader.glsl.c` in git side by side)

Both options are useful in different situations, and I think they also make sense in Zig projects, we "just" need to come up with the code that makes both cases most convenient to use...

In general I wouldn't want to export more functions (e.g. the registerModule function), but instead add more input options and return values to configure different scenarios...

...or if the 'adding more return values' gets too messy, maybe two functions: one for the out-of-source, and one for the in-source scenario...

...some things to think about:

- for the 'out-of-source' scenario, does it make sense to just return a ready-made module?
- for the 'in-source-tree' scenario, can we directly generate into the source tree, or do we need to go through the zig cache and then copy it from the cache into the source tree?
- for both options, how does this affect the `--watch` mode (I haven't tinkered with this yet, so I know nothing yet about watch mode)

nikuda committed:
Add self test on build, insource param

nikuda commented:
> we should optionally have a mode where the output file can be in the source tree and directly imported

Added `insource` param that defaults to `true` matching current behaviour.

> or the 'out-of-source' scenario, does it make sense to just return a ready-made module?

Hard for me to comment as I'm new to Zig world so keeping changes minimal and composable seems safer at the moment. On the other hand I'm sure it would be very convenient for end-user to get ready made modules. For dev side it's a fair bit to think about approaches right now and whether it will cater to all of them - or whether it will close off some things previously possible.

> for the 'in-source-tree' scenario, can we directly generate into the source tree, or do we need to go through the zig cache and then copy it from the cache into the source tree?

Looks like Zig build provides convenience functions for this exact scenario with `addUpdateSourceFiles` and `addCopyFileToSource` which register the compiled file to be outputted to source.

> for both options, how does this affect the --watch mode (I haven't tinkered with this yet, so I know nothing yet about watch mode)

With last commit https://github.com/floooh/sokol-tools-bin/pull/11/commits/99d497dbcdfdb59a59db828a22e858bbafa7c784 both in-source and out-source options are now supported with `--watch`, though `addCopyFileToSource` is not triggering properly yet - still investigating this part as I had to implement local compilation first through new `shdc_dir` param. You can try locally with `zig build test --watch` and updating `testdata/triangle.glsl` should trigger a recompile.

nikuda committed:
Test step, more precise arch handling, getShdcSubPath error

nikuda committed:
Make sure to return addUpdateSourceFiles as final if insource=true

nikuda commented:
Ok I think it's all working as intended now with https://github.com/floooh/sokol-tools-bin/pull/11/commits/2291667fdecd131e48c12b7f683b27c5a7be7ab1 - just needed to make sure to return the last step of any step dependency chain within `compile` fn. This means that end user toggling `insource` flag will not require any other changes or dependency setup.

Please give it a go with your workflow and see how you go. On this branch you can now run `zig build test --watch`, make changes to `testdata/triangle.glsl` and see `testdata/triangle.glsl.zig` rebuilt on save.

nikuda committed:
Reduce size of test case output

nikuda commented on build.zig:
```diff
@@ -78,26 +101,40 @@ fn formatToString(f: Format) []const u8 {
-        return dep_shdc.path(sub_path);
-    } else {
-        return error.ShdcUnsupportedPlatform;
+pub fn getShdcSubPath() error{ShdcUnsupportedPlatform}![]const u8 {
```
This might be a useful export as before accidentally finding `compile` is exported here I was recreating this fn to use with `b.addSystemCommand` directly. This still might be useful in some cases, i.e currently `shdc_dir` is a relative path someone might want to do an absolute path or other things.

floooh commented:
Okidoki, thanks for the changes. I'm currently deep in the sokol-gfx resource view update, but I'll try to find some time playing with the PR soon-ish!

floooh commented on build.zig:
```diff
@@ -133,11 +170,23 @@ fn optsToArgs(opts: Options, b: *Build, tool_path: Build.LazyPath) ![]const []co
 
-pub fn build(b: *Build) void {
-    _ = b;
+pub fn build(b: *Build) !void {
+    const input_path = "testdata/triangle.glsl";
+    const output_path = "testdata/triangle.glsl.zig";
+
+    const result = try compile(b, .{
+        .shdc_dir = "./",
+        .input = b.path(input_path),
+        .output = output_path,
+        .slang = .{
+            .glsl430 = true,
+        },
+        .reflection = true,
+    });
+
+    const test_step = b.step("test", "Test sokol-shdc compilation");
+    test_step.dependOn(result.step);
```
Not sure about this part (and the shader test files) tbh... I'd rather use the sokol-zig repo to check if all shader compilation scenarios are working (maybe I'll change one of the samples to use the out-of-source / module approach).

floooh replied:
...on second though... I need to think more about this... please don't remove this stuff from the PR :)

nikuda replied:
Few options come to mind if you don't want to have some sample / test data here. I think keeping `sokol-tools-bin` free of any code might be best, as this is essentially a binary delivery repo. 

First option is to move all Zig code currently in `sokol-tools-bin/build.zig` to `sokol-tools`, it already has `test/*.glsl` files there that can be used.
* Link 'pure' `sokol-tools-bin` as a git submodule at `sokol-tools/bin`. 
* Then add support both for compiled from source and using the above bin option
* If separation is wanted like `sokol` and `sokol-zig` then `sokol-tools-zig` can be created.

Option 2 might be to create new `sokol-examples` repo, pulled in by `sokol-zig`, `sokol-tools` and `sokol-tools-bin`.

I think also `sokol-tools` could be renamed to `sokol-shdc` or similar. and `sokol-tools-bin` to `sokol-shdc-bin` - it seems they're of large enough size / concern separation if they're already not in a mono repo.

I do think you want to have access to test files as it's much easier to have this close `compile` fn and test it in repo.

nikuda replied:
Another option in terms of bin delivery is to setup GitHub Actions so that artefacts can show up on Releases page. This should eliminate the need for `sokol-tools-bin`. On free plans that's 20 jobs per month, there doesn't seem to be a time limit on how long a job can take so might work out? Otherwise a [self hosted runner](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners) might be an option if demands are too great or you don't want to be tied into Github.

floooh replied:
> First option is to move all Zig code currently in sokol-tools-bin/build.zig to sokol-tools, it already has test/*.glsl files there that can be used.

I don't want to mix sokol-tools and sokol-tools-bin like that. With git submodules, sokol-tools is massive. Also apart from that the Zig package manager doesn't support submodules, so having build.zig in sokol-tools and have the binaries linked in via a submodule wouldn't work.

> Another option in terms of bin delivery is to setup GitHub Actions so that artefacts can show up on Releases page. This should eliminate the need for sokol-tools-bin.

The deploy job already has artifacts (https://github.com/floooh/sokol-tools/actions/runs/15165905627), but currently they're shortlived and only used to pass the executables between jobs, and there's an artifact missing which contains the executables for all platforms. That could be trivially fixed though.

I definitely need the delivery via git though, so that can't go away.

floooh commented:
I'm experimenting with the PR in sokol-zig with this change:

https://github.com/floooh/sokol-zig/compare/master...nikuda-shdc-test

A couple of notes:

- I think we should just remove the `.run` from the result. For some reason my dependency setup no longer worked when I tried to use the `result.run` (e.g. the shader compiler wasn't called when the output file didn't exist or the input file had changed), but when rewriting my code to `.step` it worked.
- Providing an `options.output` path with `options.insource = false` is quite confusing since the output file only lives in the `.zig-cache` - I'll try to think of something more intuitive.
- the `options.input` path should probably also be converted to a regular string slice for consistency - since the top-level project Builder object is passed into the compile function any lazy paths can just be constructed inside the function

...maybe the more intuitive solution *is* actually to have two functions:

```zig
// compile to Zig module, output source is in Zig cache only
// NOTE: no .output option
const module = shdc.createModule(b, .{
    .shdc_dep = ...,
    .input = input_path,
    .slang = .{ .glsl430 = true },
});

// compile to Zig source, output source path is relative to build.zig,
const step = shdc.createSourceFile(b, .{
    .shdc_dep = ...,
    .input = input_path,
    .output = output_path,
    .slang = .{ .glsl430 = true },
});
```

...both functions would call into a common helper function which does the `b.addSystemCommand` to invoke sokol-shdc, but then the result would be used differently, `createModule()` would wrap the output file in the zig cache into a module object, while `createSourceFile()` creates and returns the `b.addUpdateSourceFile` step.

Let me know what you think and if you want to update the PR like this, otherwise I can just do those changes myself when I get around to it :)

PS: the function signatures should probably be changed like this so that the same Options object can be used for both:

```
fn createModule(b: *Build, module_name: []const u8, opts: Options) !*Module
fn createSourceFile(b: *Build, output_path: []const u8, opts: Options) !*Step
```

...and then remove `.output` and `.insource` from Options, and turn `.input` and `.tmp_path` into regular `[] const u8` (care must be taken though that those are converted into proper 'top-level-project' paths in the optsToArgs() function, e.g. something like `b.path(...).getPath(b)` - which looks very weird though, I bet there must be a better way).

nikuda commented:
> I think we should just remove the .run from the result. For some reason my dependency setup no longer worked when I tried to use the result.run (e.g. the shader compiler wasn't called when the output file didn't exist or the input file had changed), but when rewriting my code to .step it worked.

This should be fine, Run is a subtype of Step so what is lost the ability to chain further args onto return of `compile` but looks like that is controlled tightly through `compile` params anwway. Did it fall to be called with `zig build` or fail to be re-called with `zig build --watch`?

> Providing an options.output path with options.insource = false is quite confusing since the output file only lives in the .zig-cache - I'll try to think of something more intuitive.

I think the `output` can be made optional, ie return `null` with `insource` branch. I didn't do this initially as didn' think it made much difference which path you returned with `insource` branch, and returning the cache version seemed closer to `truth` as another option might be to put the file in source, then also upload it somewhere (predicting the future here I know, but still in cache version is the 'true' source, in source version is the copy of it).

New return type would look like
```
pub const Result = struct {
    step: *Build.Step,
    output: ?Build.LazyPath,
};
```

> the options.input path should probably also be converted to a regular string slice for consistency - since the top-level project Builder object is passed into the compile function any lazy paths can just be constructed inside the function

Will do.

> maybe the more intuitive solution is actually to have two functions

Sound good.

nikuda committed:
Switch over from insource param to createSourceFile fn

nikuda commented:
Ok I think it's up to scratch now with https://github.com/floooh/sokol-tools-bin/pull/11/commits/c713bba2934f2ef29dd89e89f0b1751a8501a121, fixed up the types as well, it's a bit neater now overall.

With `createModule` implementation,  because it depends on sokol module being there to inject into generated `.glsl.zig` files, it might be more suited to have it at `sokol-zig` level? It's not as 'plug and play' if the user has to supply `sokol_module` into it. Anyway this is what a local implementation would look like, I can insert it into this PR if you prefer to keep it here after all. Can't add a `zig build test-module` step locally because of missing sokol module of course.

```
pub fn createModule(
    b: *Build,
    opts: Options,
    sokol_module: *Build.Module,
    module_name: []const u8,
) !*Build.Module {
    const output_path = try compile(b, opts);
    const shader_module = b.addModule(module_name, .{ .root_source_file = output_path });
    shader_module.addImport("sokol", sokol_module);
    return shader_module;
}
```
Edit: On second thought I added `createModule`, maybe `sokol_module` injection is not that off since the returning module needs to be wired to `root_module` so it's still not entirely plug and play.

nikuda committed:
Add createModule fn

nikuda committed:
Switch b, opts for optsToArgs sig

nikuda committed:
Change run, copy_step to const

nikuda committed:
Reduce test data size

floooh commented:
Ah shoot, I didn't think of the sokol module dependency in the generated output source file... but yeah, I think passing the sokol module into the createModule function is the best solution, since that way it's not possible to "do it wrong".

floooh commented:
Ok, I tested the `createSourceFile()` function in sokol-zig and it's working nicely, I would change one last thing though (already tested and that works too):

I think the return type `!*Build.Step.UpdateSourceFiles` of the createSourceFile function is too specific, the caller isn't really interested (nor should know) the specific step type. I would change the return type to `!*Build.Step` like this:

```zig
pub fn createSourceFile(b: *Build, opts: Options) !*Build.Step {
    const output_path = try compile(b, opts);
    const copy_step = b.addUpdateSourceFiles();
    copy_step.addCopyFileToSource(output_path, opts.output);
    return &copy_step.step;
}
```

...then my wrapper function in sokol-zig can simply look like this (it already returned a `*Build.Step` before:

```zig
fn buildExampleShader(b: *Build, example: Example) !?*Build.Step {
    if (!example.has_shader) {
        return null;
    }
    const shaders_dir = "examples/shaders/";
    return shdc.createSourceFile(b, .{
        .shdc_dep = b.dependency("shdc", .{}),
        .input = b.fmt("{s}{s}.glsl", .{ shaders_dir, example.name });
        .output = b.fmt("{s}{s}.glsl.zig", .{ shaders_dir, example.name });
        .slang = .{ ... }
        },
    });
}
```

...I still need to experiment with the module variant and also with `--watch` but with the above change I'm happy with the source-file variant :)

nikuda committed:
Return Build.Step type form createSourceFile

nikuda commented:
Changed to ` !*Build.Step`, much nicer again https://github.com/floooh/sokol-tools-bin/pull/11/commits/33bf06497ab3bc55b51413bf8e5e2cb076899abc

nikuda commented:
Next logical thing might be to think about a wrapper for this behaviour:

```zig
const shaders_dir = "examples/shaders/";
return shdc.createSourceFile(b, .{
    .shdc_dep = b.dependency("shdc", .{}),
    .input = b.fmt("{s}{s}.glsl", .{ shaders_dir, example.name });
    .output = b.fmt("{s}{s}.glsl.zig", .{ shaders_dir, example.name });
    .slang = .{ ... }
    },
});
```

ie. `createShaderFilesFromDir` (bad name) - but essentially a fn that you point to a directory and all `.glsl` file in there will be transpiled in-source. Module version of that fn might be that you get a "shaders" module back with file imports. ie 

```zig
const shd = @import("shaders").triangle;
// or easier
const shd = @import("shaders/triangle");
```

floooh commented:
Yeah, bundling multiple shaders into the same module sounds like a good idea. Let's focus first in getting this PR in (I still need to play with the module and --watch stuff), then we can think about the 'whole directory' approach :)

floooh commented:
Ok, I tested the module approach, and one small change request there:

Please change the argument order like this so that the simple args come before the `Options` args:

```zig
pub fn createModule(
    b: *Build,
    module_name: []const u8,
    sokol_module: *Build.Module,
    opts: Options,
) !*Build.Module {
```

I tested in my pacman.zig project, creating the shader module now looks like this:

```zig
fn addShaderModule(b: *Build, dep_sokol: *Build.Dependency) !*Build.Module {
    return sokol.shdc.createModule(b, "shader", dep_sokol.module("sokol"), .{
        .shdc_dep = dep_sokol.builder.dependency("shdc", .{}),
        .input = "src/shader.glsl",
        .output = "shader.zig",
        .slang = .{
            .glsl410 = true,
            .glsl300es = true,
            .hlsl4 = true,
            .metal_macos = true,
            .wgsl = true,
        },
    });
}
```

...and it's integrated with the top-level module like this:

```zig
const mod_pacman = b.createModule(.{
    .root_source_file = b.path("src/pacman.zig"),
    .target = target,
    .optimize = optimize,
    .imports = &.{
        .{ .name = "sokol", .module = dep_sokol.module("sokol") },
        .{ .name = "shader", .module = try addShaderModule(b, dep_sokol) },
    },
});
```

...this actually saves quite a bit of code, since the shader module doesn't require a dependency setup and it also only happens once (previously I had duplicated code in the native vs wasm code paths.

I could also delete the generated `shader.zig` file from the source repository.

I'm starting to think that the module approach might be the only one we need, but let's keep it at the 'dual approach' for now...

I've also been thinking of only providing the `sokol_dep` and derive everything from there (since the `shdc_dep` can be looked up from the `sokol_dep` via `sokol_dep.builder.dependency("shdc", .{})` and the sokol module via `dep_sokol.module("sokol")`, e.g. Options would look like this:

```zig
pub const Options = struct {
    sokol_dep: ?*Build.Dependency,
    input: []const u8,
    output: []const u8,
    ...
};
```

...this would simplify the call in pacman.zig's build.zig like this:

```zig
    return sokol.shdc.createModule(b, "shader", .{
        .sokol_dep = dep_sokol,
        .input = "src/shader.glsl",
        .output = "shader.zig",
        .slang = .{
            .glsl410 = true,
            .glsl300es = true,
            .hlsl4 = true,
            .metal_macos = true,
            .wgsl = true,
        },
    });
```

...this also means though that the test would need to be removed from sokol-tools-bin/build.zig.

But let's merge the MR first, I might tinker around with it later.

Please only do that one change request to change the argument order, and then I can merge :)

PS: or if you're fine with removing the test and like the idea of only providing `sokol_dep` then go ahead and do that change too, your decision :)

nikuda committed:
Shift createModule params around

nikuda commented:
> thinking of only providing the sokol_dep and derive everything from there

This looks good with me too 👍 We lose the ability to test within the repo which is just a convience. If you have a way to link this repo with `sokol-zig` then nothing is really lost. But if code is copied manually from this repo to test in `sokol-zig` then some fragility is introduced there.

Either way another way can be thought of I'm sure that makes testing easier.

floooh commented:
On further thought - the `sokol_dep` idea wouldn't work from within the sokol-zig project... I'll just merge the current state, fixup my Zig projects, and any future simplications can go into separate PRs :)

floooh commented:
PS: I'm also thinking about moving the code back into the sokol-zig build.zig - the main reason it was moved into sokol-tools-bin was because other sokol bindings also used the Zig build system, but that's no longer the case.

floooh commented:
Ok merged. Many thanks!

floooh commented:
Resulting sokol-zig PR: https://github.com/floooh/sokol-zig/pull/121

floooh commented:
...and as module in pacman.zig: https://github.com/floooh/pacman.zig/pull/24

floooh commented:
...and another example of the shader module approach in chipz:

https://github.com/floooh/chipz/commit/47199ddccc0bc95322227c4071c1d5139b62015c

nikuda commented:
> thinking about moving the code back into the sokol-zig build.zig

This makes a lot of sense to me 👍 Think it'll be easier to find there, in an ideal scenario `sokol-tools-bin` wouldn't exist at all right.
